go/obscuronode/enclave/enclave.go

```diff
@@ -326,71 +346,88 @@ func (e *enclaveImpl) produceRollup(b *types.Block, bs *obscurocore.BlockState)
+
+				// the speculative execution has been processing on top of the wrong parent - due to failure in gossip or publishing to L1
+				// or speculative execution is disabled
+				speculativeExecutionSucceeded = speculativeRollup.found && (speculativeRollup.r.Hash() == bs.HeadRollup)
+
+				if !speculativeExecutionSucceeded && speculativeRollup.r != nil {
+					nodecommon.LogWithID(e.nodeShortID, "Recalculate. speculative=r_%d(%d), published=r_%d(%d)",
+						obscurocommon.ShortHash(speculativeRollup.r.Hash()),
+						speculativeRollup.r.Header.Number,
+						obscurocommon.ShortHash(bs.HeadRollup),
+						headRollup.Header.Number)
+					if e.statsCollector != nil {
+						e.statsCollector.L2Recalc(e.nodeID)
+					}
+				}
 			}
+	*/
+
+	successfulTransactions := make([]nodecommon.L2Tx, 0)
+	// if !speculativeExecutionSucceeded {
+	// In case the speculative execution thread has not succeeded in producing a valid rollup
+	// we have to create a new one from the mempool transactions
+	newRollupHeader = obscurocore.NewHeader(&bs.HeadRollup, headRollup.Header.Number+1, e.nodeID)
+	newRollupTxs = currentTxs(headRollup, e.mempool.FetchMempoolTxs(), e.storage)
+	newRollupState = e.storage.CreateStateDB(bs.HeadRollup)
+	receipts := evm.ExecuteTransactions(newRollupTxs, newRollupState, newRollupHeader, e.storage, e.chainID)
+	// todo - only transactions that fail because of the nonce should be excluded
+	for _, tx := range newRollupTxs {
+		_, f := receipts[tx.Hash()]
+		if f {
+			successfulTransactions = append(successfulTransactions, tx)
+		} else {
+			log.Info(">   Agg%d: Excluding transaction %d", obscurocommon.ShortAddress(e.nodeID), obscurocommon.ShortHash(tx.Hash()))
 		}
 	}
 
-	if !speculativeExecutionSucceeded {
-		// In case the speculative execution thread has not succeeded in producing a valid rollup
-		// we have to create a new one from the mempool transactions
-		newRollupHeader = obscurocore.NewHeader(&bs.HeadRollup, headRollup.Header.Number+1, e.nodeID)
-		newRollupTxs = currentTxs(headRollup, e.mempool.FetchMempoolTxs(), e.storage)
-
-		newRollupState = e.storage.CreateStateDB(bs.HeadRollup)
-		executeTransactions(newRollupTxs, newRollupState, newRollupHeader)
-	}
-
 	// always process deposits last, either on top of the rollup produced speculatively or the newly created rollup
 	// process deposits from the proof of the parent to the current block (which is the proof of the new rollup)
 	proof := e.blockResolver.Proof(headRollup)
-	depositTxs := extractDeposits(proof, b, e.blockResolver, e.erc20ContractLib)
-	executeTransactions(depositTxs, newRollupState, newRollupHeader)
+	depositTxs := extractDeposits(proof, b, e.blockResolver, e.erc20ContractLib, newRollupState, e.chainID)
+	depositReceipts := evm.ExecuteTransactions(depositTxs, newRollupState, newRollupHeader, e.storage, e.chainID)
+	for _, tx := range depositTxs {
+		if depositReceipts[tx.Hash()] == nil {
+			panic("Should not happen")
+		}
+	}
```

Comment:
Should this be done upstream in the evm ?

Reply:
Umm, not really. The evm allows transactions to fail. But these deposit transactions are synthetic, so they should not fail.

Reply:
I see. Since we're doing the evm facade I suppose I'd expect the facade to handle these synthetic txs.
I wouldn't expect it to return an array of empty txs (or a few), maybe it could return an error 
It just feels weird that we're checking the result here instead of doing it in the facade ?

Reply:
the evm facade's responsibility is to execute the transactions supplied to it against the database it's been given.
If it fails, it returns that.
It should be the layers above to decide what to do in case the execution failed. Why should the execution layer know why you're executing something?